incubator/virtualcluster/pkg/controller/virtualcluster/virtualcluster_controller.go

```diff
@@ -68,13 +111,266 @@ var _ reconcile.Reconciler = &ReconcileVirtualcluster{}
 // ReconcileVirtualcluster reconciles a Virtualcluster object
 type ReconcileVirtualcluster struct {
 	client.Client
-	scheme *runtime.Scheme
+	scheme          *runtime.Scheme
+	clusterVersions map[string]*tenancyv1alpha1.ClusterVersion
+	virtualClusters map[string]*tenancyv1alpha1.Virtualcluster
+}
+
+func (r *ReconcileVirtualcluster) checkCRKind(request *reconcile.Request) (runtime.Object, CRKind, error) {
```

Comment:
You cannot prevent vc and clusterversion from having same name which breaks this method. I suggest to use another reconcile loop to reconcile for clusterversion.

Reply:
when a new vc cr is created, we have to refer to existing cluster versions. This cannot be achieved by using two separate controllers. In other words, without a client set, we can’t query apiserver when we want to get all available clusterversions.